test/orange/surf/detail/QuarticSolver.test.cc

```diff
@@ -0,0 +1,219 @@
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/surf/detail/QuarticSolver.test.cc
+//---------------------------------------------------------------------------//
+#include "corecel/Types.hh"
+#include "corecel/cont/Array.hh"
+#include "corecel/cont/Range.hh"
+#include "orange/surf/detail/FerrariSolver.hh"
+
+#include "celeritas_test.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+namespace test
+{
+//---------------------------------------------------------------------------//
+/*
+ * Template test harness for arbitrary quartic solvers
+ */
+template<typename MyQuarticSolver>
+class QuarticSolverTest : public ::celeritas::test::Test
+{
+  public:
+    using Intersections = Array<real_type, 4>;
+    using Coeffs4 = Array<real_type, 4>;
+    using Coeffs5 = Array<real_type, 5>;
+
+    QuarticSolverTest() {}
+
+    template<class MQS = MyQuarticSolver>
+    Intersections get_roots(Coeffs5 const& abcde)
+    {
+        MQS solve_quartic;  // Default tolerance
+        return solve_quartic(abcde);
+    }
+
+    Intersections fill_inf(std::initializer_list<real_type> const& up_to_four)
+    {
+        Intersections all_roots;
+        all_roots.fill(no_intersection());
+        int i = 0;
+        for (auto val : up_to_four)
+        {
+            all_roots[i] = val;
+            i++;
+        }
+        return all_roots;
+    }
+
+    void expect_softeq_list(Intersections const& expected,
+                            Intersections const& actual)
+    {
+        ASSERT_EQ(expected.size(), actual.size());
+        for (auto i : range(expected.size()))
+        {
+            EXPECT_SOFT_EQ(expected[i], actual[i]);
+        }
+    }
+
+    void expect_surface_roots_from_coeffs(
+        std::initializer_list<real_type> const& expected, Coeffs4 const& abcd)
+    {
+        auto [a, b, c, d] = abcd;
+        auto x = this->get_roots({a, b, c, d, 0});
+        this->expect_softeq_list(fill_inf(expected), x);
+    }
+
+    void expect_nonsurface_roots_from_coeffs(
+        std::initializer_list<real_type> const& expected, Coeffs5 const& abcde)
+    {
+        auto x = this->get_roots(abcde);
+        this->expect_softeq_list(fill_inf(expected), x);
+    }
+};
+
+//---------------------------------------------------------------------------//
+
+using QuarticSolvers = ::testing::Types<FerrariSolver>;
+TYPED_TEST_SUITE(QuarticSolverTest, QuarticSolvers, );
+
+/*
+ * Test cases with all non-zero roots, i.e., the ray does not start on or close
+ * to the surface
+ */
+
+TYPED_TEST(QuarticSolverTest, no_roots)
+{
+    // x^4 + 2*x^3 - 2.999998*x^2 - 3.999998*x + 4.000005000001
+    // Four complex roots 1+-0.001i, -2+-0.001i
+    {
+        this->expect_nonsurface_roots_from_coeffs(
+            {}, {1, 2, -2.999998, -3.999998, 4.000005000001});
```

Comment:
I recommend changing `fill_inf` to a free function `make_roots`, declaring your solver locally, and writing all your tests as:
```c++
TypeParam solve{}; // Solver with default tolerance

// Four complex roots...
EXPECT_VEC_EQ(make_roots({}), 
        solve({1, 2, -2.999998, -3.999998, 4.000005000001}));

// Two negative real roots...
EXPECT_VEC_EQ(make_roots({}), ...);

// etc
```

Your output will be clearer and your test smaller.